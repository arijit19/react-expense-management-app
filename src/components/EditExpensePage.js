import React from 'react';
import {connect} from 'react-redux';
import { withRouter} from 'react-router-dom';

import ExpenseForm from './ExpenseForm.js';
import {StartRemoveExpense, StartEditExpense} from '../Redux/Actions/ExpenseActions.js'


export class EditExpensePage extends React.Component {
   
    onSubmit = (expense)=> {
        // console.log(expense);
        
        this.props.startEditExpense(this.props.match.params.id.toString().slice(1),expense);
        this.props.history.push('/');
    };

    onClick = ()=> {
        // console.log(this.props);
        
        this.props.startRemoveExpense({id: this.props.match.params.id.toString().slice(1)});
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h3 className='page-header__title'>Editing expence</h3>
                    </div>
                </div>
                <div className='content-container'>
                    <ExpenseForm 
                        expense = {this.props.expense}
                        onSubmit= {this.onSubmit}   
                    />
                    <button onClick={this.onClick} className='button button--secondary'>Remove Expense</button>
                </div>     
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
      expense: state.expenses.find((expense) => expense.id === props.match.params.id.toString().slice(1))
    }
  };

const mapDispatchToProps = (dispatch)=> {
    return {
        startEditExpense: (id,expense)=> dispatch(StartEditExpense(id,expense)),
        startRemoveExpense: (data)=> dispatch(StartRemoveExpense(data))
    }
}


export default connect(mapStateToProps,mapDispatchToProps)(withRouter(EditExpensePage));