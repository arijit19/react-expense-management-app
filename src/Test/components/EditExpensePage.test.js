import React from 'react';
import {shallow} from 'enzyme';

import {EditExpensePage} from '../../components/EditExpensePage.js'
import expenses from '../fixtures/Expense.js'

let wrapper, match, editExpenseSpy, removeExpenseSpy, historySpy;

beforeEach(()=>{
     match = {
        params: {
            id: `:${expenses[0].id}`
        }
    }
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = { push: jest.fn()};
    wrapper = shallow(<EditExpensePage 
        match={match} 
        expense={expenses[0]} 
        editExpense={editExpenseSpy} 
        removeExpense={removeExpenseSpy} 
        history={historySpy}
        />);
})

test('Testing Editing Expences component render',()=>{
    
    expect(wrapper).toMatchSnapshot();
})

test('Testing Editing Expences component for editing items',()=>{
    const expense = {
        ...expenses[0],
        note: 'New Note',
    }
   wrapper.find('ExpenseForm').prop('onSubmit')(expense);
   expect(historySpy.push).toHaveBeenLastCalledWith('/');
   expect(editExpenseSpy).toHaveBeenLastCalledWith(expense.id,expense);
})

test('Testing Editing Expences component for removing items', ()=>{
    wrapper.find('button').prop('onClick')();
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({id: expenses[0].id})
})