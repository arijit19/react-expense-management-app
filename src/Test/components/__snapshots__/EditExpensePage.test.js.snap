// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Editing Expences component render 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h3
        className="page-header__title"
      >
        Editing expence
      </h3>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 20000,
          "createdAt": 0,
          "description": "Rent",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
